storage/backup/backup-cli/src/coordinators/restore.rs

```diff
@@ -88,110 +89,308 @@ impl RestoreCoordinator {
-        //      snapshot version.
-        // And it does so in a resume-able way.
-
+    /// Support two modes
+    /// 1. restore to target version when two_phase_restore is false
+    /// 2. restore a DB with all data ranging from start_version to target_version when two phase restore is true
+    /// The overall flow is as follows:
+    /// The first phase is restore till the tree snapshot before the target version. It includes the following work
+    /// a. start from the first transaction of loaded chunk, save the txn accumualator, and apply transactions till the first kv snapshot. We don't restore KV here since we don't know the storage usage as this point
+    /// b. restore kv snapshot before ledger history start version, which also restore the storage usage.
+    /// c. save transaction and retore KV till the tree_snapshot before target version
+    /// The second phase is restore the tree snapshot and replay txns till the target version
+    /// a. restore the tree snapshot
+    /// b. replay the txn till the target version
+    ///
+    /// we are support the resume from any point when the restore is interrupted.
+    async fn run_impl(self) -> Result<()> {
         if self.replay_all {
             bail!("--replay--all not supported in this version.");
         }
 
+        info!("This tool only guarantees resume from previous in-progress restore. \
+        If you want to restore a new DB, please either specify a new target db dir or delete previous in-progress DB in the target db dir.");
+
         let metadata_view = metadata::cache::sync_and_load(
             &self.metadata_cache_opt,
             Arc::clone(&self.storage),
             self.global_opt.concurrent_downloads,
         )
         .await?;
 
-        let next_txn_version = self
+        let target_version = self.global_opt.target_version;
+        COORDINATOR_TARGET_VERSION.set(target_version as i64);
+
+        // calculate the start_version and replay_version
+        let max_txn_ver = metadata_view
+            .max_transaction_version()?
+            .ok_or_else(|| anyhow!("No transaction backup found."))?;
+        let lhs = self.ledger_history_start_version();
+
+        let latest_tree_version = self
+            .global_opt
+            .run_mode
+            .get_state_snapshot_before(Version::MAX);
+        let tree_completed = {
+            match latest_tree_version {
+                Some((ver, _)) => self
+                    .global_opt
+                    .run_mode
+                    .get_state_snapshot_before(ver)
+                    .is_some(),
+                None => false,
+            }
+        };
+
+        let mut db_next_version = self
             .global_opt
             .run_mode
             .get_next_expected_transaction_version()?;
-        if next_txn_version != 0 {
-            // DB is already in workable state
+
+        let kv_snapshot = if db_next_version == 0 {
```

Comment:
in progress version is none and db_next_version to decide if kv_snapshot is some